File: GRPCClient.cpp
/*******************************************************************************
 * MICROSOFT CONFIDENTIAL & PROPRIETARY
 *
 * Copyright (c) 2019 Microsoft Corp.
 * All Rights Reserved.
 ******************************************************************************/

#include "GRPCClient.hpp"

#include <memory>
#include <string>

#include <grpcpp/grpcpp.h>

#include "SwitchML.grpc.pb.h"

// using grpc::Channel;
// using grpc::ClientContext;
// using grpc::Status;
// using SwitchML::RDMAConnectRequest;
// using SwitchML::RDMAConnectResponse;
// using SwitchML::RDMAServer;


GRPCClient::GRPCClient(std::shared_ptr<grpc::Channel> channel)
  : stub(SwitchML::RDMAServer::NewStub(channel))
{ }

void GRPCClient::RDMAConnect(const SwitchML::RDMAConnectRequest & request,
                             SwitchML::RDMAConnectResponse * response) {
  grpc::ClientContext context;
  grpc::Status status = stub->RDMAConnect(&context, request, response);
  if (!status.ok()) {
    std::cerr << "Error contacting coordinator: "
              << status.error_code() << ": "
              << status.error_message()
              << std::endl;
    exit(1);
  }
}

std::ostream & operator<<(std::ostream & s, const SwitchML::RDMAConnectRequest & r) {
  s << "<RDMAConnectRequest" << std::hex
    << " rank=" << r.my_rank()
    << " size=" << r.job_size()
    << " mac=0x" << r.mac()
    << " ipv4=0x" << r.ipv4()
    << " rkey=0x" << r.rkey();
  for (int i = 0; i < r.qpns_size(); ++i) {
    s << " qpn=0x" << r.qpns(i) << " psn=0x" << r.psns(i);
  }
  return s << ">";
}

std::ostream & operator<<(std::ostream & s, const SwitchML::RDMAConnectResponse & r) {
  s << "<RDMAConnectResponse" << std::hex;
  for (int i = 0; i < r.macs_size(); ++i) {
    s << " mac=0x" << r.macs(i)
      << " ipv4=0x" << r.ipv4s(i)
      << " rkey=0x" << r.rkeys(i);
  }
  for (int i = 0; i < r.qpns_size(); ++i) {
    s << " qpn=0x" << r.qpns(i) << " psn=0x" << r.psns(i);
  }
  return s << ">";
}

File: py/GRPCClient.py
######################################################
#   Copyright (C) Microsoft. All rights reserved.    #
######################################################

import logging
from pprint import pprint, pformat

import grpc

import SwitchML_pb2
import SwitchML_pb2_grpc

def run_test():
    #with grpc.insecure_channel('127.0.0.1:50099') as channel:
    with grpc.insecure_channel('localhost:50099') as channel:
        stub = SwitchML_pb2_grpc.RDMAServerStub(channel)
        print("Sending request")

        job_size = 1
        for rank in range(job_size):
            response = stub.RDMAConnect(SwitchML_pb2.RDMAConnectRequest(
                job_id = 12345,
                my_rank = rank,
                job_size = job_size,
                mac = 0,
                ipv4 = 0,
                rkey = 12345,
                message_size = 1024,
                qpns = [1, 2, 3, 4, 5],
                psns = [6, 7, 8, 9, 0]
            ))
            print("GRPC client received:\n{}".format(pformat(response)))

if __name__ == '__main__':
    logging.basicConfig()
    run_test()
            

